File: Fortran Version/Bending Waveguide/bending_waveguide.f90
program fdtd_bending_waveguide_simulation_optimized
  use, intrinsic :: iso_fortran_env, only: dp => real64
  implicit none

  !-----------------------------------------------------------------
  ! Parameters and variable declarations.
  !-----------------------------------------------------------------
  integer, parameter :: output_interval = 1  ! Frame output frequency.
  integer :: nx, ny, nsteps, i, j, n, i_source, j_source
  real(dp) :: c, dx, dy, dt, t, cfl, Cx, Cy, source_val, omega
  real(dp) :: x_min, x_max, y_min, y_max
  real(dp) :: channel_halfwidth
  real(dp), allocatable, target :: field_old(:,:), field(:,:), field_new(:,:)
  real(dp), pointer :: Ez_old(:,:), Ez(:,:), Ez_new(:,:)
  real(dp), allocatable :: x(:), y(:)
  real(dp) :: pi, alpha
  real(dp) :: start_time, end_time
  integer :: file_unit
  real(dp), allocatable :: center(:), bottom_bound(:), top_bound(:)
  real(dp), pointer :: tmp(:,:)  ! Temporary pointer for swapping

  !-----------------------------------------------------------------
  ! Define physical constants and grid parameters.
  !-----------------------------------------------------------------
  pi = acos(-1.0_dp)
  c = 3.0e8_dp           ! Speed of light in m/s

  ! Domain: x from 0 to 100 m; y from -50 to 50 m.
  x_min = 0.0_dp
  x_max = 100.0_dp
  y_min = -50.0_dp
  y_max = 50.0_dp

  ! Grid spacing and grid dimensions.
  dx = 0.1_dp
  dy = 0.1_dp
  nx = int((x_max - x_min) / dx + 0.5_dp) + 1
  ny = int((y_max - y_min) / dy + 0.5_dp) + 1

  ! Total number of time steps.
  nsteps = 2000

  ! CFL condition with safety factor.
  cfl = 0.999999999999999_dp
  dt = cfl * dx / (c * sqrt(2.0_dp))
  print *, "Adaptive dt based on CFL condition: ", dt, " s"

  !-----------------------------------------------------------------
  ! Allocate field arrays and associate pointers for pointer swapping.
  ! The target attribute is needed for pointer association.
  !-----------------------------------------------------------------
  allocate(field_old(nx, ny), field(nx, ny), field_new(nx, ny))
  Ez_old => field_old
  Ez     => field
  Ez_new => field_new

  ! Allocate coordinate arrays.
  allocate(x(nx), y(ny))
  do i = 1, nx
     x(i) = x_min + (i - 1) * dx
  end do
  do j = 1, ny
     y(j) = y_min + (j - 1) * dy
  end do

  ! Initialize all fields to zero.
  Ez_old = 0.0_dp
  Ez     = 0.0_dp
  Ez_new = 0.0_dp

  ! Finite-difference coefficients.
  Cx = (c * dt / dx)**2
  Cy = (c * dt / dy)**2

  !-----------------------------------------------------------------
  ! Define the bending waveguide geometry.
  !-----------------------------------------------------------------
  channel_halfwidth = 10.0_dp  ! Half–width of the channel (in meters).

  ! Precompute the channel center and PEC boundaries for each x.
  allocate(center(nx), bottom_bound(nx), top_bound(nx))
  do i = 1, nx
     center(i)      = 5.0_dp * sin(2.0_dp * pi * x(i) / 50.0_dp)
     bottom_bound(i) = center(i) - channel_halfwidth
     top_bound(i)    = center(i) + channel_halfwidth
  end do

  !-----------------------------------------------------------------
  ! Set up the source.
  ! Place the source near x = 5 m at the local channel center.
  !-----------------------------------------------------------------
  i_source = int((5.0_dp - x_min) / dx + 0.5_dp) + 1
  j_source = int((center(i_source) - y_min) / dy + 0.5_dp) + 1

  ! Source frequency: 10 MHz (angular frequency).
  omega = 2.0_dp * pi * 1.0e7_dp

  !-----------------------------------------------------------------
  ! Open a file to save all frame data.
  !-----------------------------------------------------------------
  open(newunit = file_unit, file = "bending_waveguide_fdtd.dat", status = "replace", &
       action = "write", form = "formatted")

  !-----------------------------------------------------------------
  ! Start timing the simulation.
  !-----------------------------------------------------------------
  call CPU_TIME(start_time)

  !-----------------------------------------------------------------
  ! Main time-stepping loop.
  !-----------------------------------------------------------------
  do n = 1, nsteps
     t = n * dt

     !--------------------------------------------------------------
     ! Finite-difference update for interior points.
     ! The loops are re-ordered (j outer, i inner) to improve memory
     ! access (Fortran arrays are column-major) and are parallelized.
     !--------------------------------------------------------------
!$omp parallel do collapse(2) default(shared) private(i,j)
     do j = 2, ny - 1
        do i = 2, nx - 1
           Ez_new(i, j) = 2.0_dp * Ez(i, j) - Ez_old(i, j) + &
                           Cx * (Ez(i + 1, j) - 2.0_dp * Ez(i, j) + Ez(i - 1, j)) + &
                           Cy * (Ez(i, j + 1) - 2.0_dp * Ez(i, j) + Ez(i, j - 1))
        end do
     end do
!$omp end parallel do

     !--------------------------------------------------------------
     ! Add the sinusoidal source term.
     !--------------------------------------------------------------
     source_val = sin(omega * t)
     Ez_new(i_source, j_source) = Ez_new(i_source, j_source) + source_val

     !--------------------------------------------------------------
     ! Enforce PEC (perfect electric conductor) conditions outside
     ! the bending waveguide.
     ! Since the channel boundaries (bottom_bound and top_bound) were
     ! precomputed, we simply test each point.
     !--------------------------------------------------------------
!$omp parallel do default(shared) private(j)
     do i = 1, nx
        do j = 1, ny
           if (y(j) < bottom_bound(i) .or. y(j) > top_bound(i)) then
              Ez_new(i, j) = 0.0_dp
           end if
        end do
     end do
!$omp end parallel do

     !--------------------------------------------------------------
     ! Apply Mur absorbing boundary conditions on the left and right.
     !--------------------------------------------------------------
!$omp parallel do default(shared) private(j)
     alpha = (c * dt - dx) / (c * dt + dx)
     do j = 2, ny - 1
        Ez_new(1, j)  = Ez(2, j) + alpha * (Ez_new(2, j) - Ez(1, j))
        Ez_new(nx, j) = Ez(nx - 1, j) + alpha * (Ez_new(nx - 1, j) - Ez(nx, j))
     end do
!$omp end parallel do

     !--------------------------------------------------------------
     ! Save the current frame to file.
     ! (Here we write every output_interval time step.)
     !--------------------------------------------------------------
     if (mod(n, output_interval) == 0) then
        write(file_unit, *) "Frame ", n, " Time ", t, " NX ", nx, " NY ", ny
        do j = 1, ny
           write(file_unit, *) (Ez_new(i, j), i = 1, nx)
        end do
        write(file_unit, *)  ! Blank line to separate frames.
     end if

     !--------------------------------------------------------------
     ! Swap pointers (rotate the roles of old, current, and new arrays)
     ! instead of copying arrays. This avoids a full-array copy.
     !--------------------------------------------------------------
     tmp     => Ez_old
     Ez_old  => Ez
     Ez      => Ez_new
     Ez_new  => tmp
  end do

  !-----------------------------------------------------------------
  ! End timing the simulation.
  !-----------------------------------------------------------------
  call CPU_TIME(end_time)
  close(file_unit)
  
  print *, "Simulation completed. All frames saved in bending_waveguide_fdtd.dat."
  print *, "Elapsed CPU time: ", end_time - start_time, " seconds."
  
end program fdtd_bending_waveguide_simulation_optimized

File: Waveguides/straight_waveguide/bending_waveguide.jl
    using Plots

    function run_fdtd_bending_waveguide_simulation()
        # ---------------------------
        # Simulation and Grid Setup
        # ---------------------------
        # Use the real speed of light (in m/s)
        c = 3.0e8                        # Wave speed (m/s)
        # Domain: x from 0 to 100 m; y from -10 to 10 m (wider domain to accommodate the bend)
        x_min, x_max = 0.0, 100.0
        y_min, y_max = -50.0, 50.0

        # Grid spacing (in meters)
        dx = 0.1
        dy = 0.1
        nx = Int(round((x_max - x_min) / dx)) + 1
        ny = Int(round((y_max - y_min) / dy)) + 1

        # Create coordinate arrays for plotting
        x = range(x_min, x_max, length=nx)
        y = range(y_min, y_max, length=ny)

        # ---------------------------
        # CFL Condition: Adaptive dt
        # ---------------------------
        # With physical c, the time step becomes very small.
        cfl = 0.999999999999999  # CFL safety factor (≤ 1.0)
        dt = cfl * dx / (c * sqrt(2))
        println("Adaptive dt based on CFL condition: ", dt, " s")

        nsteps = 2000  # Total number of time steps

        # ---------------------------
        # Field Initialization
        # ---------------------------
        Ez_old = zeros(Float64, nx, ny)  # Field at time step n-1
        Ez     = zeros(Float64, nx, ny)  # Field at time step n
        Ez_new = zeros(Float64, nx, ny)  # Field at time step n+1

        # Finite-difference coefficients (for second derivatives):
        Cx = (c * dt / dx)^2
        Cy = (c * dt / dy)^2

        # ---------------------------
        # Define Bending Waveguide Geometry
        # ---------------------------
        # Define the channel half–width and a bending centerline.
        channel_halfwidth = 10.0  # in meters
        center_line(x_val) = 5.0 * sin(2π * x_val / 50)  # centerline oscillates with amplitude 2 m

        # ---------------------------
        # Source Parameters
        # ---------------------------
        # Place the source near x = 5 m, at the channel center.
        source_x = 5.0
        source_center = center_line(source_x)
        i_source = Int(round((source_x - x_min) / dx)) + 1
        j_source = Int(round((source_center - y_min) / dy)) + 1

        # Set the source frequency to 10 MHz (so that the wavelength ~ 30 m)
        # (You may adjust this along with grid resolution in a real simulation.)
        omega = 2π * 1.0e7  # Angular frequency in rad/s

        # ---------------------------
        # Animation Loop
        # ---------------------------
        anim = @animate for n = 1:nsteps
            t = n * dt

            # Update interior grid points (finite-difference update)
            @inbounds for i in 2:nx-1
                for j in 2:ny-1
                    Ez_new[i, j] = 2 * Ez[i, j] - Ez_old[i, j] +
                        Cx * (Ez[i+1, j] - 2 * Ez[i, j] + Ez[i-1, j]) +
                        Cy * (Ez[i, j+1] - 2 * Ez[i, j] + Ez[i, j-1])
                end
            end

            # Add the sinusoidal source term at the designated grid point.
            source_val = sin(omega * t)
            Ez_new[i_source, j_source] += source_val

            # ---------------------------
            # Enforce PEC Conditions Outside the Bending Waveguide
            # ---------------------------
            # For each x, compute the local channel boundaries and zero the field outside.
            for i in 1:nx
                c_val = center_line(x[i])
                bottom_bound = c_val - channel_halfwidth
                top_bound = c_val + channel_halfwidth
                for j in 1:ny
                    if y[j] < bottom_bound || y[j] > top_bound
                        Ez_new[i, j] = 0.0
                    end
                end
            end

            # ---------------------------
            # Mur Boundary Condition for x (left and right edges)
            # ---------------------------
            alpha = (c * dt - dx) / (c * dt + dx)
            for j in 2:ny-1
                # Left boundary (i = 1)
                Ez_new[1, j] = Ez[2, j] + alpha * (Ez_new[2, j] - Ez[1, j])
                # Right boundary (i = nx)
                Ez_new[nx, j] = Ez[nx-1, j] + alpha * (Ez_new[nx-1, j] - Ez[nx, j])
            end

            # ---------------------------
            # Plot the Field
            # ---------------------------
            # Transpose the field so that rows correspond to y.
            p = heatmap(x, y, Ez', aspect_ratio=1,
                clims=(-0.1, 0.1), color=:RdBu,
                xlabel="x (m)", ylabel="y (m)",
                title="2D FDTD Bending Waveguide Simulation, t = $(round(t, sigdigits=3)) s",
                xlims=(x_min, x_max), ylims=(y_min, y_max))
            # Overlay the bending waveguide walls
            bottom_curve = [center_line(x_val) - channel_halfwidth for x_val in x]
            top_curve = [center_line(x_val) + channel_halfwidth for x_val in x]
            plot!(p, x, bottom_curve, lw=2, color=:red, label="Wall")
            plot!(p, x, top_curve, lw=2, color=:red, label="")

            # ---------------------------
            # Advance to Next Time Step
            # ---------------------------
            Ez_old .= Ez
            Ez     .= Ez_new

            p  # Return the plot for the current frame.
        end

        # Save the animation as a GIF (adjust fps as desired)
        gif(anim, "fdtd_bending_waveguide.gif", fps=30)
    end

    # Run the simulation
    run_fdtd_bending_waveguide_simulation()
